### PR TITLE
[DSA] Integrate Q8KV8 FP8 Sparse MLA Prefill into the DSA Backend (DeepSeek-V3.2)

### PR DESCRIPTION
**Pull request by JackChuang**
## Summary
This PR integrates the SM90 Q8KV8 FP8 sparse-MLA prefill kernel from PR-1 (#25751) into SGLang's **DSA prefill backend** (`dsa_backend.py`), so DeepSeek-V3.2 serving runs **native FP8 query × FP8 KV** sparse attention during prefill end-to-end. It is opt-in via `--dsa-prefill-backend flashmla_sparse_q8` (the `--nsa-prefill-backend` alias still works; requires `fp8_e4m3` KV; the bf16 baseline remains `flashmla_sparse`), and adds the concurrency/correctness fixes the kernel needed under real multi-rank load. Decode is unchanged (`flashmla_kv`).

- **Performance:** beats the bf16 (`flashmla_sparse`) baseline on every shape — single-request **+4.6–7.7%** (peak +7.7% @16k), saturated throughput **+1.8–2.7%** (H200, TP8/EP8/DP8, both arms fp8 KV; medians over repeated A/B rounds).
- **Accuracy:** **q8 matches q16 across 3.6k–65k context** — identical 1319-question unseen gsm8k at every rung, per-rung **|Δ| ≤ 0.004** (pooled −0.002, negligible); both arms flat at **0.95–0.96** with no context-length degradation. On **LongBench-v2** (thinking-ON, unique 10k–147k-token documents — the kernel-heaviest workload): **0.602 vs 0.616, a gap within the protocol's sampling noise (Δ −0.014 ± 0.015)**.

## Motivation

PR-1 (#25751) delivered the Q8KV8 kernel as a standalone, independently reviewable block. Running sparse-MLA prefill as native FP8 — rather than dequantizing the fp8 KV back to bf16 and using the bf16 kernel — avoids a paged-fp8→bf16 round-trip, halves KV-side memory traffic, and unlocks FP8 tensor-core throughput, turning PR-1's kernel speedup into an end-to-end serving win.

## Context

Part of roadmap #25746. PR-1 (#25751) = the standalone SM90 kernel + unit tests + microbenchmark. This PR adds, on top of that kernel: the **DSA-backend integration** (`dsa_backend.py`); **additional kernel-correctness coverage** (concurrency/correctness fixes that only surfaced under real DP8 end-to-end load); and **end-to-end accuracy + performance validation** supplementing PR-1's kernel-level microbenchmark/tests.

## Modifications

| file | role |
|---|---|
| `dsa_backend.py` | integration + `flashmla_sparse_q8` routing (`_forward_flashmla_sparse_q8kv8`), dispatch, RAGGED gate, startup validation + decode guard |
| `triton_ops/cache_ops.py` (re-exported via `utils.py`) | fused Triton q-prep: concat(q_nope, q_rope) + bf16→fp8 cast (`concat_and_cast_q_fp8_pad`) |
| `dsa/dequant_k_cache.py` | fused gather + dequant + requant of paged fp8 KV (`gather_dequant_requant_fp8_paged`; no fp8→bf16→fp8 round-trip) |
| `server_args.py` | `flashmla_sparse_q8` in `DSA_CHOICES` |
| `kernel.cuh` | concurrency/correctness fixes on the PR-1 kernel (race, deadlock, stale-P, masked-softmax, proxy fence) |
| `test/registered/jit/…_sm90.py` | +6 e2e-discovered corner-case gates folded into the PR-1 unit tests (21 = 15 + 6) |

**Key points:**
- **Explicit backend selection:** `--dsa-prefill-backend flashmla_sparse_q8` is a first-class DSA prefill impl (no hidden env; `--nsa-prefill-backend` is a deprecated alias); `flashmla_sparse` keeps the bf16 path. Always FP8 + SM90-only, both validated at startup with explicit errors (fail-fast); rejected as a `--dsa-decode-backend` (prefill-only).
- **Native FP8 KV (no round-trip):** the paged fp8 KV is gathered, dequantized per-group, and re-quantized per-tensor in one fused kernel. Identity per-tensor scales — a raw bf16→fp8 cast is accurate on real DeepSeek-V3 magnitudes, so no dynamic rescaling.
- **Fused FP8 q-prep:** concat q_nope/q_rope and cast bf16→fp8 in one pass — worth **~+3.5% c=1 throughput / ~+1.7% saturated** end-to-end vs a separate concat + cast (measured at il=16k and 32k). The Triton kernel supports power-of-two head counts only (guarded at the call site, with a generic concat + cast fallback; DeepSeek's 128 heads and all its TP splits qualify).
- **Correctness fixes from e2e:** full-topk dispatch (−1 sentinels clamped to distinct zero pad rows ⇒ uniform control flow across DP ranks — a data-dependent loop bound had deadlocked the cross-rank collective at long context), plus a `fence.proxy.async` determinism fix (orders the V-transpose STSM writes before the consumer WGMMA reads).

## Usage

```bash
python -m sglang.launch_server --model-path <DeepSeek-V3.2> \
  --dsa-prefill-backend flashmla_sparse_q8 \   # native FP8 sparse prefill (this PR)
  --dsa-decode-backend flashmla_kv \
  --kv-cache-dtype fp8_e4m3 \
  --tp-size 8 --ep-size 8 --dp-size 8 --enable-dp-attention --trust-remote-code
```

`flashmla_sparse_q8` is always FP8 (requires `--kv-cache-dtype fp8_e4m3`) and SM90-only — both asserted at startup. For the bf16 baseline use `--dsa-prefill-backend flashmla_sparse` (the only change needed to A/B). Prompts whose prefill context ≤ `index_topk` (2048) take the dense one-shot path, where both backends are identical.

### Performance (H200, TP8/EP8/DP8, both arms fp8 KV)

q8 wins every rung of all four tables below: at the default chunked-prefill size (32768), single-request latency **+4.6–7.7%** (peak +7.7% @ il=16k) and saturated throughput **+1.8–2.7%**; at the secondary chunk size (16384), single-request **+2.6–6.9%** and saturated **+1.6–2.7%** (il ≤ 2k excluded: dense path, both arms identical). The edge is largest at low load and shrinks toward saturation — profiling shows an overlap/makespan effect (un-overlapped per-layer all-reduces at c=1 amplify a small compute saving), not a compute-share effect. Numbers are medians over repeated A/B rounds; each row's Δ varies by only ≈ 0.3% across rounds, so the deltas sit well outside run-to-run noise. All settings are upstream defaults.

**chunk32768 — c=1 single-request latency:**

| il | q8 tps | q16 tps | Δtps | q8 TTFT | q16 TTFT | ΔTTFT |
|---|--:|--:|--:|--:|--:|--:|
| 3.5k | 7,310 | 6,987 | +4.6% | 475 ms | 497 ms | +4.4% |
| 8k | 7,566 | 7,029 | +7.6% | 1,052 ms | 1,132 ms | +7.1% |
| 16k | 7,379 | 6,855 | +7.7% | 2,165 ms | 2,336 ms | +7.3% |
| 32k | 6,948 | 6,464 | +7.5% | 4,607 ms | 4,952 ms | +7.0% |
| 64k | 6,123 | 5,728 | +6.9% | 10,434 ms | 11,149 ms | +6.4% |
| 120k | 5,042 | 4,758 | +6.0% | 23,818 ms | 25,239 ms | +5.6% |
| 160k | 4,453 | 4,227 | +5.3% | 35,927 ms | 37,902 ms | +5.2% |

**chunk32768 — saturated throughput (sat_c per il):**

| il (sat_c) | q8 tps | q16 tps | Δtps | q8 TTFT | q16 TTFT | ΔTTFT |
|---|--:|--:|--:|--:|--:|--:|
| 3.5k (c64) | 27,745 | 27,247 | +1.8% | 8,050 ms | 8,205 ms | +1.9% |
| 8k (c32) | 27,311 | 26,625 | +2.6% | 9,398 ms | 9,631 ms | +2.4% |
| 16k (c16) | 26,574 | 25,927 | +2.5% | 9,691 ms | 9,940 ms | +2.5% |
| 32k (c16) | 25,508 | 24,888 | +2.5% | 19,988 ms | 20,492 ms | +2.5% |
| 64k (c16) | 23,502 | 22,902 | +2.6% | 42,368 ms | 43,446 ms | +2.5% |
| 120k (c16) | 21,082 | 20,521 | +2.7% | 68,005 ms | 69,923 ms | +2.7% |
| 160k (c16) | 19,503 | 18,998 | +2.7% | 98,076 ms | 100,587 ms | +2.5% |

**chunk16384 — c=1 single-request latency:**

| il | q8 tps | q16 tps | Δtps | q8 TTFT | q16 TTFT | ΔTTFT |
|---|--:|--:|--:|--:|--:|--:|
| 3.5k | 8,410 | 8,200 | +2.6% | 414 ms | 423 ms | +2.1% |
| 8k | 7,719 | 7,241 | +6.6% | 1,036 ms | 1,100 ms | +5.8% |
| 16k | 7,202 | 6,739 | +6.9% | 2,218 ms | 2,373 ms | +6.5% |
| 32k | 6,624 | 6,196 | +6.9% | 4,828 ms | 5,157 ms | +6.4% |
| 64k | 5,777 | 5,437 | +6.2% | 11,070 ms | 11,765 ms | +5.9% |
| 120k | 4,780 | 4,528 | +5.6% | 25,123 ms | 26,524 ms | +5.3% |
| 160k | 4,222 | 4,025 | +4.9% | 37,932 ms | 39,801 ms | +4.7% |

**chunk16384 — saturated throughput (sat_c per il):**

| il (sat_c) | q8 tps | q16 tps | Δtps | q8 TTFT | q16 TTFT | ΔTTFT |
|---|--:|--:|--:|--:|--:|--:|
| 3.5k (c64) | 26,950 | 26,537 | +1.6% | 8,361 ms | 8,499 ms | +1.6% |
| 8k (c32) | 26,412 | 25,816 | +2.3% | 9,729 ms | 9,945 ms | +2.2% |
| 16k (c16) | 26,057 | 25,449 | +2.4% | 9,945 ms | 10,186 ms | +2.4% |
| 32k (c16) | 24,926 | 24,382 | +2.2% | 20,157 ms | 20,609 ms | +2.2% |
| 64k (c16) | 23,141 | 22,604 | +2.4% | 43,249 ms | 44,315 ms | +2.4% |
| 120k (c16) | 20,601 | 20,260 | +1.7% | 69,148 ms | 70,379 ms | +1.7% |
| 160k (c16) | 19,240 | 18,731 | +2.7% | 98,132 ms | 101,094 ms | +2.9% |

**Where the advantage comes from (profile decomposition).** Both arms serve the identical il=64k workload (c=1 and c=16) while sglang's built-in torch profiler captures a full 8-rank trace per arm (`/start_profile` → bench → `/stop_profile`); GPU kernel time is bucketed by kernel name and averaged across ranks. Δ = q16 − q8 = GPU-busy time q8 saves per rank over the capture window:

| Δ = q16 − q8 (mean per-rank GPU-busy, ms) @ il=64k | c=1 (latency) | c=16 (saturated) |
|---|--:|--:|
| attention prefill kernel (direct delta; q8 kernel −10.9%) | +154 | +564 |
| indexer / topk | +76 | +89 |
| **NCCL (barrier amplification)** | **+1,876** | +1,046 |
| q-prep / cast | +26 | +146 |
| MoE / GEMM / other | +23 | +116 |
| **total** | **+2,156 (+6.4% of q16 busy)** | **+1,962 (+2.4%)** |

The totals independently match the end-to-end tables (+6.4% ↔ the 64k c=1 ΔTTFT +6.4%; +2.4% ↔ saturated +2.5%). Reading: at saturation the win converges to the direct kernel delta; at c=1 the per-layer all-reduces are un-overlapped and stall on the slowest rank, so q8 reaching each barrier sooner turns a 154 ms kernel saving into 1,876 ms less collective spin — an **overlap/makespan** effect, which is why the single-request advantage (+5–8%) exceeds the saturated one (+2–3%).

### Correctness

sglang-standard pytest `tests/test_sparse_mla_q8kv8_prefill_sm90.py` — **21/21 passed**. The PR-1 suite plus 6 new corner-case gates that real e2e serving exposed:

| Group | Cases | Coverage |
|---|--:|---|
| Original PR-1 suite | 15 | correctness (d_qk 512/576), sink/topk_length, corner shapes, precision, buffer ownership |
| NEW — masked −1-sentinel | 2 | few-valid rows (ctx < topk) vs a −inf fp32 reference; cos 0.9998 |
| NEW — s_q envelope | 3 | s_q 2048/4096/6144 first-band NaN watch; cos 0.9999 |
| NEW — large S_KV | 1 | S_KV=65536 / large gathered index; cos 0.9999 |

#### gsm8k

End-to-end, both arms fp8 KV, chunk32768, greedy decoding (temperature 0), uniform methodology — context length is the only variable.

**One methodology, one variable.** Every row is the SAME test: gsm8k, **all 1319 test questions, identical at every rung, all unseen** (few-shot examples come from the gsm8k *train* split, so no evaluated question's answer ever appears in the prompt); completion API, temperature 0, same stop strings and scoring. The **only variable is context length** (the few-shot count). Accuracy = fraction correct (higher = better); Δ = q8 − q16 (negative = q8 lower); σ = same-config run-to-run standard deviation.

| context (tokens) | method | q8 | q16 | Δ (q8−q16) | σ (q8 / q16) |
|---|---|--:|--:|--:|--:|
| 3.6k (3,574) | 20-shot | 0.956 | 0.958 | −0.002 | 0.001 / 0.001 |
| 6.9k (6,943) | 40-shot | 0.952 | 0.955 | −0.004 | 0.002 / 0.001 |
| 13k (13,466) | 80-shot | 0.956 | 0.958 | −0.002 | 0.001 / 0.000 |
| 20k (20,074) | 120-shot | 0.955 | 0.954 | +0.000 | 0.003 / 0.004 |
| 26k (26,383) | 160-shot | 0.958 | 0.958 | +0.000 | 0.002 / 0.004 |
| 40k (40,364) | 240-shot | 0.955 | 0.958 | −0.003 | 0.001 / 0.003 |
| 65k (65,375) | 390-shot | 0.955 | 0.958 | −0.003 | 0.002 / 0.001 |

**q8 matches q16 across 3.6k–65k context**: per-rung **|Δ| ≤ 0.004** (all within the σ ≈ 0.001–0.004 run-to-run noise floor), pooled Δ = **−0.002** — a very small quantization cost in the expected direction, negligible in magnitude. Both arms are flat at **0.95–0.96 with no context-length degradation**.

Note: few-shot examples are drawn from the *train* split (lm-eval convention; one-line variant of the stock gsm8k bench, script archived) so every evaluated question is unseen at every rung.

#### LongBench-v2

End-to-end, both arms fp8 KV, chunk32768; thinking-ON at the official protocol — temperature 0.6 / top-p 0.95, max-tokens 16,384 (sampling is required for long CoT and keeps absolutes leaderboard-comparable; the single-run sampling noise floor is ≈ ±0.027 at N=334, so we report means over repeated runs); harness = sglang's built-in `run_eval --eval-name longbench_v2`.

The adversarial accuracy test for a sparse-*prefill* kernel is one where every request runs a full sparse prefill over a long, *unique* context — no shared prefix, no dense-path dilution — and the answer depends on that prefill being right. LongBench-v2 is exactly that: after the 2,300–147,000-token filter, all 334 scored four-way multiple-choice questions (exact letter-match scoring, no judge) carry a unique 10k–147k-token document (median 51k), so each request invokes the kernel under test **183–2,196 times** (prefill chunks × 61 MLA layers), and the up-to-16k-token thinking chain then amplifies any prefill numerics into the final answer.

| arm | mean score | σ |
|---|--:|--:|
| q8 (`flashmla_sparse_q8`) | **0.602** | 0.012 |
| q16 (`flashmla_sparse`) | **0.616** | 0.023 |

Δ = **−0.014 ± 0.015 (SE of the mean difference) — within the protocol's sampling noise** (< 1σ; same-config repeat runs spread by up to 0.045, more than the mean gap itself): no measurable fp8 accuracy cost on the workload with maximal kernel exposure. Both arms were measured on the exact submitted revision.

## Known Upstream Issue (not introduced by this PR)

- **DSA fused top-k v2 (paged) crashes with an illegal memory access on ~160k-token prefill** — affects `flashmla_sparse` and `flashmla_sparse_q8` identically (shared indexer/top-k path, upstream of both prefill kernels); il ≤ 120k is unaffected. Workaround: `SGLANG_OPT_USE_TOPK_V2=0` — with it the il = 160k benches pass and reproduce the table's deltas. Tracked upstream: **<ISSUE LINK — TBD>**.

070826 update: Seems like the 'DSA fused top-k v2' issue is solved by #30512 

## Next
- Will release GLM5.2 support and optimization soon, (perf gain upto 2x~4x compared to dpsk-v3.2).
- I also noticed if I optimize the e2e shared path, q8 kernel path can also bring more performance gain. So those optimization will be release along with the next GLM5.2 PR to further improve the dpsk-v3.2's  performance.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29509668482](https://github.com/sgl-project/sglang/actions/runs/29509668482)
Latest PR Test (Extra): :white_check_mark: [Run #29509761289](https://github.com/sgl-project/sglang/actions/runs/29509761289)

### CODE REVIEW

**Review by BBuf — APPROVED** [review]
LGTM.
